[Step1] 문자열 덧셈 계산기 코드 리뷰 요청드립니다.

ohtaeg opened the pull request:
안녕하세요 리뷰어님.
기능별로 커밋을 했었어야 했는데 실수로 한번에 커밋을 해버렸습니다.
다시 돌아가서 고치고나서 커밋하기에는 오래걸릴것 같아 이번 한번만 양해 부탁드립니다. 😭 
다음부터는 기능별로 커밋하도록 노력하겠습니다.  😁 

저의 좋지않은 코드를 리뷰 해주셔서 감사합니다.
클린코드를 작성 할 수 있도록 열심히 하겠습니다 ❗️ 
잘 부탁 드립니다 :)

ohtaeg committed:
feat(string calculator) : 문자열 덧셈 계산기 구현

- 빈 문자열 or null 입력시 0을 반환 한다.
- 빈 문자열 or null 외 다른 숫자 문자열 입력시 0을 반환하지 않는다.
- 숫자만 입력했을경우 해당 숫자를 반환 한다.
- 숫자 사이에 구분자 쉼표(,) 또는 콜론(:)이 있을경우 분리후 숫자의 합을 반환 한다.
- 커스텀 구분자로 문자열 분리후 숫자의 합을 반환 한다.

ohtaeg committed:
test(string calculator) : 문자열 덧셈 계산기 테스트 코드 작성

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/Number.java:
```diff
@@ -0,0 +1,39 @@
+    private int number;
+    public static final int ZERO = 0;
+    public static final int ONE = 1;
+    public static final int TWO = 2;
```
클래스는 상수, 멤버 변수, 생성자, 메서드 순으로 작성합니다.

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/Number.java:
```diff
@@ -0,0 +1,39 @@
+    public static final int ONE = 1;
+    public static final int TWO = 2;
+
+    public Number() {
```
`Number`를 생성할 때 완전한 상태로 생성할 수 있게 변경하면 어떤 장점이 있을까요?

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/Number.java:
```diff
@@ -0,0 +1,39 @@
+            throw new RuntimeException(
+                    String.format(Constants.RUNTIME_EXCEPTION_MESSAGE, number));
+        }
+        this.number += number;
```
좋은 객체란 무엇일까요? 아래의 글을 읽고 느낀 점을 전체 코드에 반영해 보면 어떨까요?
https://codingnuri.com/seven-virtues-of-good-object

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.regex.Pattern;
+
+public class StringCalculator {
+    private final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile("//(.)\n(.*)");
```
미리 `Pattern`으로 만들어 사용 👍

ohtaeg replied:
#37 
Number 객체를 인스턴스화 하는 곳에서 객체 생성 후 값을 set 해주지 않을 실수를 유발할 수 있기 때문에
해당 Number 객체를 생성시 값도 set하여 완전한 상태로 객체 생성을 할 수 있도록 해주는것이
사용자 실수를 줄여주는 장점이 있는 것 같아 소스를 변경해보았습니다.

알려주신 글을 읽고나서, 우선 클래스 설계를 잘못했다는 것을 느꼈습니다.
각 객체는 자신의 특성이 아닌 준수하고 있는 계약(협력)을 위해 책임과 역할이 분명 해야 하는데 그렇지 않았고, 또한 고유함을 유지하도록 캡슐화가 되어 있어야 하는데 그렇지도 않았습니다.
아직은 많이 부족하지만 소스를 수정 해보았습니다.

바쁘신 와중에도 리뷰 해주셔서 감사합니다.
또 부족한 부분이 있다면 피드백 부탁 드립니다.
열심히 하겠습니다.

wotjd243 reviewed (changes requested):
깔끔한 코드 구현이 좋았습니다. 👍 
간단한 생각거리를 남겼으니 코드에 반영해 주셔도 좋습니다.
진행 과정에서 궁금한 점은 댓글이나 Slack을 통해 물어보세요.

ohtaeg committed:
refactor (string calculator) : 피드백 반영

- Number 생성시 완전한 상태로 생성할 수 있게 변경

ohtaeg committed:
test (string calculator) : Number 클래스 테스트 코드 작성

ohtaeg committed:
refactor(string calculator) : Number 객체 생성시 생성자 대신 정적 팩토리 메서드 사용하도록 변경

- new Number(value) -> Number.intValueOf(value);
- static 메서드를 통해 객체 생성 캡슐화를 위함
- int로 Number 객체를 생성한다는것을 쉽게 읽히고 알리게 하기 위함

ohtaeg committed:
test (string calculator) : Number 객체 생성시 생성자 대신 정적 팩토리 메서드 사용하도록  테스트 코드 변경

ohtaeg committed:
refactor(string calculator) : 정적멤버 변수 제거

ohtaeg committed:
refactor (string calculator) : Number 클래스의 불필요한 상수와 StringCalculator 클래스의 역할에 맞지 않는
separate() 함수와 delimiter 상수들 Delimiter 클래스로 따로 분리

ohtaeg committed:
test (string calculator) : Delimiter 클래스 문자열 분리 테스트 코드 작성

ohtaeg commented:
a

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/Number.java:
```diff
@@ -0,0 +1,41 @@
+        if (isNegative(number)) {
+            throw new RuntimeException(String.format(Constants.RUNTIME_EXCEPTION_MESSAGE, number));
+        }
+        this.value += number;
```
아래의 코드로 개선해 보면 어떨까요?

```suggestion
        this.value = number;
```

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,27 @@
+            numbers.add(Number.intValueOf(text));
+        }
+        return numbers.stream()
+                      .mapToInt(Number::getValue)
```
`Number`에 메시지를 던지도록 구조를 바꿔 보면 어떨까요? 객체에 메시지를 보내 데이터를 가지는 객체가 일하도록 해보세요.

wotjd243 reviewed (changes requested):
피드백을 전부 잘 반영해 주셨습니다. 💯 
간단한 생각거리를 남겼으니 코드에 반영해 주셔도 좋습니다.
다음 리뷰 요청 때는 Merge 하도록 하겠습니다.

ohtaeg committed:
refactor(string calculator) : Numbers 클래스 추가

ohtaeg committed:
test (string calculator) : Numbers 클래스 테스트 코드 추가

ohtaeg committed:
refactor (string calculator) : Numbers 클래스 정적 팩터리 메서드 추가 및 sum() 메서드 추가

- StringCalculator 클래스에서 계산이 아닌 메시지를 통해 Numbers클래스에서 데이터를 다루도록 수정.

ohtaeg committed:
refactor (string calculator) : sum() 메서드 제거

ohtaeg committed:
test (string calculator) : Numbers 클래스 sum() 메서드 테스트 코드 추가

ohtaeg commented:
안녕하세요 제이슨님!
보내주신 피드백을 통해 수정 해보았습니다.

https://github.com/next-step/ddd-kitchenpos/pull/37#discussion_r350572087
현재 StringCalculator 클래스의 sum() 메서드는 반복문을 통해 Number 객체를 생성후
List에 add후 바로 데이터를 꺼내 로직을 처리하도록 구현이 되어있었던 부분을
객체에 메시지를 보내 데이터를 가진 객체가 일을 할 수 있도록 Numbers 클래스를 새로 생성하여 구현하였습니다.

Numbers 클래스를 새로 생성한 이유는 다음과 같은 장점들이 있는것 같아 추가하였습니다.
- 기존에 존재하던 List 데이터 집합을 캡슐화하여 은닉.
- 메시지 기반으로 객체에 일을 할 수 있도록 역할을 좀 더 확실하게 부여하여 객체간 상호작용 유도
- 가독성 개선

미흡하거나 또다른 피드백을 주신다면 다음부터 설계 할때 반영 하도록 하겠습니다.
감사합니다.

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/Numbers.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.List;
+import java.util.Objects;
+
+public final class Numbers {
```
이미 일급 컬렉션의 개념을 알고 계신 것 같아요. 👍 아래의 글을 읽고 일급 컬렉션을 내 것으로 확실히 완성시켜 보세요.
https://jojoldu.tistory.com/412

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/Numbers.java:
```diff
@@ -0,0 +1,43 @@
+        for (String text : separatedText) {
+            numbers.add(Number.intValueOf(text));
+        }
+        return new Numbers(Collections.unmodifiableList(numbers));
```
새로운 컬렉션에 담아 외부에 의한 변조를 막는 것은 습관처럼 사용하는 것이 좋습니다. 👍 아래의 글을 읽고 얕은 복사와 깊은 복사에 대해 공부해 보세요.
https://isooo.github.io/etc/2019/01/11/shallowCopy-vs-deepCopy.html

wotjd243 replied:
아래의 코드처럼 주 생성자(primary constructor)에서 새로운 컬렉션에 담으면 어떨까요?

```java
private Numbers(List<Number> numbers) {
    this.numbers = Collections.unmodifiableList(numbers);
}
```

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/Numbers.java:
```diff
@@ -0,0 +1,43 @@
+    public Number sum() {
+        int sum = 0;
+        for (Number number : numbers) {
+            sum += number.getValue();
```
`Number`까지 메시지를 던지면 어떨까요?

wotjd243 reviewed (changes requested):
코드 작성에 심혈을 기울인 것이 느껴져 기분 좋게 리뷰를 진행하였습니다. 💯 
간단한 생각거리를 남겼으니 코드에 반영해 주셔도 좋습니다.
다음 리뷰 요청 때는 Merge 하도록 하겠습니다.

ohtaeg committed:
refactor (string calculator) : 변경 불가 컬렉 만드는 행위를 생성자에서 하도록 수정

ohtaeg committed:
refactor (kitchen pos) :  Number 객체에 더하라는 메시지 던지도록 add() 메서드 추가

ohtaeg committed:
test (kitchen pos) :  Number 클래스의 Number 객체 더하는 add() 메서드 로직 테스트 코드 추가

ohtaeg committed:
refactor (string calculator) : fianl 키워드 추가

ohtaeg commented:
안녕하세요 제이슨님 😄 
https://github.com/next-step/ddd-kitchenpos/pull/37#discussion_r351599745, https://github.com/next-step/ddd-kitchenpos/pull/37#discussion_r351600587 
알려주신 내용을 토대로 좋은 공부 하였습니다. 감사합니다!
아래는 보내주신 피드백을 통해 수정 해보았습니다.

https://github.com/next-step/ddd-kitchenpos/pull/37#discussion_r351601353
Numbers클래스의 intValuesOf() 메서드에서 현재 문자열 배열을 분리해 
Numbers 객체에 담을 수 있도록 데이터를 처리하는 행위만을 해야하는데 
변경불가 컬렉션 만드는 행위를 하도록 더 한것 같아 생성자에서 변경불가 컬렉션 만들도록 수정하였습니다.

https://github.com/next-step/ddd-kitchenpos/pull/37#discussion_r351601846
Numbres 객체의 sum() 메소드는 각 Number 객체를 합해주는 행위만 하면되는데
기존 로직을 보니 Number 객체의 데이터 갖고 오고, 더하고, 결과값을 Number 객체로 바꾸는
많은 행위를 하도록 작성 했었습니다.
그렇기에 행위 분리 및 역할에 맞도록 Number 객체에게 ‘자기 자신은 너가 더해’ 라는 메시지를 넘기도록 수정해보았습니다.

수정하다보니 매직넘버("0")가 생겨 어떻게 처리할지 고민이 생겼습니다.

<pre>
<code>java</code>
<code>public Number sum() {
        final Number sum =  Number.intValueOf("0");
        for (final Number number : numbers) {
            sum.add(number);
        }
        return sum;
    }</code>
<code>```</code>
</pre>
매직넘버를 제거하기위해 2가지 방법을 생각해 봤습니다.
1. default 생성자를 새로 작성하여 디폴트 생성자만 호출했을경우 0을 가진 Number 객체를  생성하는 방법
2. 상수화 시키는 방법

이 2가지 방법을 고려해보았습니다만 둘다 적절하지 않다고 생각했습니다. 
그 이유는 
1. 기존 설계는 Number 객체 생성시 값을 넘겨 완전한 객체로 생성되도록 유도했기때문에 디폴트 생성자를 추가하는것은 바람직 하지 않다.
2. 상수 추가할 경우 이 상수의 위치는 어디가 적절한지 판단이 되질 않았습니다.

이렇게 보니 애초에 Number 객체 생성시 String이 아닌 int값을 받도록 유도하는게
맞지 않았나 싶기도 하네요.

어떤 방법으로 접근을 해야했었는지 혹시 피드백을 주실 수 있으시면 감사하겠습니다.

제이슨님 덕분에 많이 배워가네요 !!
리뷰 정말 감사드립니다. 👍 👍 👍

ohtaeg committed:
refactor (string calculator) : Number 클래스 내부에서 원시값 변경하던 로직을 일급 컬렉션에 맞게 원시값 변경 로직 제거

ohtaeg committed:
test (string calculator) : Number 클래스sum() 메서드  테스트 코드 변경

ohtaeg commented:
안녕하세요 제이슨님 😄

알려주신 내용을 토대로 좋은 공부 하였습니다. 감사합니다!
아래는 보내주신 피드백을 통해 수정 해보았습니다.


https://github.com/next-step/ddd-kitchenpos/pull/37#discussion_r351601353
Numbers클래스의 intValuesOf() 메서드에서 현재 문자열 배열을 분리해서
Numbers 객체에 담을 수 있도록 데이터를 처리하는 행위만을 해야 하는데
변경불가 컬렉션 만드는 행위까지 추가 된 것 같아 생성자에서 변경불가 컬렉션 만들도록 수정하였습니다.

https://github.com/next-step/ddd-kitchenpos/pull/37#discussion_r351601846
Numbres 객체의 sum() 메소드는 각 Number 객체를 합해주는 행위만 하면되는데
기존 로직을 보니 Number 객체의 데이터 get하고, 더하고, 결과값을 Number 객체로 바꾸는
많은 행위를 하도록 작성 했었습니다.
그렇기에 행위 분리 및 역할에 맞도록 Number 객체에게 ‘자기 자신은 너가 더해’ 라는 메시지를 넘기도록 수정해보았습니다.

수정하다보니 매직넘버("0")가 생겨 어떻게 처리할지 고민이 생겼습니다.

public Number sum() {
    final Number sum =  Number.intValueOf("0");
    for (final Number number : numbers) {
        sum.add(number);
    }
    return sum;
}
매직넘버를 제거하기위해 2가지 방법을 생각해 봤습니다.

default 생성자를 새로 작성하여 디폴트 생성자만 호출했을경우 0을 가진 Number 객체를 생성하는 방법
상수화 시키는 방법
이 2가지 방법을 고려해보았습니다만 둘다 적절하지 않다고 생각했습니다.
그 이유는

기존 설계는 Number 객체 생성시 값을 넘겨 완전한 객체로 생성되도록 유도했기때문에 디폴트 생성자를 추가하는것은 바람직 하지 않다.
상수 추가할 경우 이 상수의 위치는 어디가 적절한지 판단이 되질 않았습니다.
이렇게 보니 애초에 Number 객체 생성시 String이 아닌 int값을 받도록 설계하는게 맞지 않았나 싶기도 하네요.

어떤 방법으로 접근을 해야 했었는지 혹은 어떻게 처리 했어야 했는지에 대해
견해를 여쭤보려 합니다.. 🙏 🙏

많이 바쁘실텐데 신경써주셔서 감사합니다.
제이슨님 덕분에 많이 배워 갑니다 !!
리뷰 정말 감사드립니다. 👍 👍 👍

ohtaeg committed:
refactor (string calculator) : 0값을 갖는 Number 상수화 처리

ohtaeg commented:
안녕하세요 제이슨님 😄

알려주신 내용을 토대로 좋은 공부 하였습니다. 감사합니다!
아래는 보내주신 피드백을 통해 수정 해보았습니다.

#37 (comment)
일급 컬렉션의 특징중 하나인 불변 컬렉션을 유지하기위해
내부 데이터를 건드리지 않도록 수정하였습니다.

#37 (comment)
Numbers클래스의 intValuesOf() 메서드에서 현재 문자열 배열을 분리해서
Numbers 객체에 담을 수 있도록 데이터를 처리하는 행위만을 해야 하는데
변경불가 컬렉션 만드는 행위까지 추가 된 것 같아 생성자에서 변경불가 컬렉션 만들도록 수정하였습니다.

#37 (comment)
Numbres 객체의 sum() 메소드는 각 Number 객체를 합해주는 행위만 하면되는데
기존 로직을 보니 Number 객체의 데이터 get하고, 더하고, 결과값을 Number 객체로 바꾸는
많은 행위를 하도록 작성 했었습니다.
그렇기에 행위 분리 및 역할에 맞도록 Number 객체에게 ‘자기 자신은 너가 더해’ 라는 메시지를 넘기도록 수정해보았습니다.

많이 바쁘실텐데 신경써주셔서 감사합니다.
제이슨님 덕분에 많이 배워 갑니다 !!
리뷰 정말 감사드립니다. 👍 👍 👍

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/Number.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.Objects;
+
+public final class Number {
```
`Number`와 같은 클래스를 VALUE OBJECT라고 부른답니다. 💯 VALUE OBJECT는 불변이 특징입니다.

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/Number.java:
```diff
@@ -0,0 +1,46 @@
+
+public final class Number {
+    public static final int ZERO_VALUE = 0;
+    public static final Number ZERO_NUMBER = new Number(ZERO_VALUE);
```
인스턴스를 미리 만들어 놓거나 새로 생성한 인스턴스를 캐싱 하여 재활용하는 식으로 불필요한 객체 생성을 피할 수 있다. 👍

wotjd243 commented on src/main/java/camp/nextstep/edu/calculator/Number.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public Number add(final Number number) {
+        return new Number(this.value + number.getValue());
```
멤버 변수에 직접 접근하는 것이 아닌 getter 메서드를 사용한 이유가 있을까요?

wotjd243 reviewed (approved):
수많은 피드백을 반영하고자 대단히 고생 많으셨습니다. 💯 
"인내는 쓰고 열매는 달다."라는 말처럼 정말 달콤한 열매를 수확했길 바랍니다.
이만 Merge 하겠습니다.
